File: pbmx_ffi/java/io/rmf/pbmx/Token.java
package io.rmf.pbmx;

import io.rmf.pbmx.ffi.LibPbmx;
import io.rmf.pbmx.ffi.RawToken;

public final class Token {
    RawToken raw;

    Token(RawToken raw) {
        this.raw = raw;
    }

    public long decode() {
        long value = LibPbmx.INSTANCE.pbmx_decode_token(this.raw.val());
        assert value != -1;
        return value;
    }

    public static Token encode(long value) {
        assert value != -1;
        return new Token(LibPbmx.INSTANCE.pbmx_encode_token(value).ref());
    }
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/Token.kt
package io.rmf.pbmx

import io.rmf.pbmx.ffi.FFI
import io.rmf.pbmx.ffi.RawToken

class Token internal constructor(internal var raw: RawToken): Comparable<Token> {

    fun decode(): Long {
        val value = FFI.pbmx_decode_token(this.raw.value())
        assert(value != -1L)
        return value
    }

    override fun equals(other: Any?): Boolean {
        return this.decode() == (other as Token).decode()
    }

    override fun hashCode(): Int {
        return this.decode().hashCode()
    }

    override fun compareTo(other: Token): Int {
        return this.decode().compareTo(other.decode())
    }

    override fun toString(): String {
        return this.decode().toString()
    }

    companion object {

        fun encode(value: Long): Token {
            assert(value != -1L)
            return Token(FFI.pbmx_encode_token(value).ref())
        }
    }
}
